DistroLauncher/algorithms.h

```diff
@@ -18,40 +18,80 @@
 
 /// Common algorithms to be used everywhere in the launcher project with style resembling the std ones.
 
-/// Returns true if [tested] starts with [end]. Null-termination is not required.
-template <typename CharT>
-// NOLINTNEXTLINE(bugprone-easily-swappable-parameters) - Parameters must have the same type.
-bool starts_with(const std::basic_string_view<CharT> tested, const std::basic_string_view<CharT> start)
+namespace algo_internal
 {
-    if (tested.size() < start.size()) {
-        return false;
+    // Tells chars appart from other types
+    template <typename T> constexpr bool is_character() noexcept
+    {
+        return std::is_same_v<T, char> || std::is_same_v<T, wchar_t> || std::is_same_v<T, signed char> ||
+               std::is_same_v<T, unsigned char> || std::is_same_v<T, char16_t> || std::is_same_v<T, char32_t>;
+        // C++20 introduces char8_t
     }
-    auto mismatch = std::mismatch(start.cbegin(), start.cend(), tested.cbegin());
-    return mismatch.first == start.cend();
-}
 
-template <typename CharT, std::size_t TestedSize, std::size_t StartSize>
-bool starts_with(CharT const (&tested)[TestedSize], CharT const (&start)[StartSize])
-{
-    return starts_with<CharT>(std::basic_string_view<CharT>{tested}, std::basic_string_view<CharT>{start});
+    /**
+     * This template converts:
+     * - T[N] into basic_string_view<T> (works for non-chars too!)
+     * - CharT* into basic_string_view<CharT>
+     * - Anything else stays the same
+     */
+    template <typename T> constexpr T const& view_adaptor(T const& value)
+    {
+        return value;
+    }
+
+    template <std::size_t Size, typename T> constexpr auto view_adaptor(const T value[Size])
```

Comment:
The array will always decay into a pointer, thus this case is never instantiated, which seems to be actually fine, otherwise this would be a bug:

```cpp
const std::string tested{"Ubuntu 22.04 LTS"};
const char sub[7] = "Ubuntu"; //6 chars + \0.
// would fail because `std::basic_string_view<const char>(sub, 7);` would forcebly include the '\0' in the view, then `mismatch.first` would be an iterator to the last char, not the one-after-last.
ASSERT_TRUE(starts_with(tested, sub));
```
If you comment out the pointer specialization you'll see errors due the '\0' being included in the string_view due the array specialization.

I think this must be removed.

Reply:
This is on purpose. `char`-like types decay to a pointer which is what we want, but non-chars don't see the declaration of the pointer version (due to the `enable_if`) and hence stay as c-arrays. I can explicitly disable it for `char`-like types, though.

Reply:
Done, now pointer decay is only enabled for chars; and c-array is only enabled for non-chars:
```c++
template<typename T>
enable_if_t<!is_character<T>(), basic_string_view<T>> view_adaptor(const T value[Size]);

template<typename CharT>
enable_if_t<is_character<CharT>(), basic_string_view<CharT>> view_adaptor(const CharT* value)
```

The need for concepts strikes once again 😓